redbpf/src/lib.rs

```diff
@@ -351,31 +356,37 @@ impl Program {
         if self.data().fd.is_some() {
             return Err(Error::ProgramAlreadyLoaded);
         }
+        // Should bind CString to local variable not to make a dangling pointer
+        // with .as_ptr() method
+        let cname = CString::new(self.data().name.clone())?;
         let clicense = CString::new(license)?;
-        let cname = CString::new(self.data_mut().name.clone())?;
-        let log_buffer: MutDataPtr =
-            unsafe { libc::malloc(mem::size_of::<i8>() * 16 * 65535) as MutDataPtr };
-        let buf_size = 64 * 65535_u32;
 
-        let fd = unsafe {
-            bpf_sys::bcc_prog_load(
-                self.to_prog_type(),
-                cname.as_ptr() as DataPtr,
-                self.data_mut().code.as_ptr(),
-                (self.data_mut().code.len() * mem::size_of::<bpf_insn>()) as i32,
-                clicense.as_ptr() as DataPtr,
-                kernel_version as u32,
-                0_i32,
-                log_buffer,
-                buf_size,
-            )
-        };
+        let mut attr = unsafe { mem::zeroed::<bpf_sys::bpf_load_program_attr>() };
 
-        if fd < 0 {
-            Err(Error::BPF)
-        } else {
-            self.data_mut().fd = Some(fd);
-            Ok(())
+        attr.prog_type = self.to_prog_type();
+        attr.expected_attach_type = 0;
+        attr.name = cname.as_ptr();
+        attr.insns = self.data().code.as_ptr();
+        attr.insns_cnt = self.data().code.len() as u64;
+        attr.license = clicense.as_ptr();
+        attr.__bindgen_anon_1.kern_version = kernel_version;
+        attr.log_level = 1;
+
+        unsafe {
+            let buf_size = 64 * 1024;
+            let log_buffer: MutDataPtr = libc::malloc(buf_size) as MutDataPtr;
```

Comment:
I'm not a huge fan of `malloc` call here, especially since we free the memory area after the call anyway.

I think a neater option would be using a `[u8; buf_size]` array, and the raw pointer for that. Is there any non-trivial issue why that wouldn't work?

Reply:
I was focusing on eliminating bcc and I tried to avoid touching anything else.
malloc has been used there and I just did not change it. ;)
OK, I will replace malloc with an array.

Reply:
`malloc` is removed and `Vec.as_mut_ptr` is called to provide log buffer.